internal/initializer/webhook_tls.go

```diff
@@ -0,0 +1,177 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initializer
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+)
+
+const (
+	errGetWebhookSecret    = "cannot get webhook tls secret"
+	errUpdateWebhookSecret = "cannot update webhook tls secret"
+	errGenerateCertificate = "cannot generate tls certificate"
+)
+
+// CertificateGenerator can return you TLS certificate valid for given domains.
+type CertificateGenerator interface {
+	Generate(domain ...string) (key []byte, crt []byte, err error)
+}
+
+// WebhookCertificateGeneratorOption is used to configure WebhookCertificateGenerator behavior.
+type WebhookCertificateGeneratorOption func(*WebhookCertificateGenerator)
+
+// WithCertificateGenerator sets the CertificateGenerator that
+// WebhookCertificateGenerator uses.
+func WithCertificateGenerator(cg CertificateGenerator) WebhookCertificateGeneratorOption {
+	return func(w *WebhookCertificateGenerator) {
+		w.certificate = cg
+	}
+}
+
+// NewWebhookCertificateGenerator returns a new WebhookCertificateGenerator.
+func NewWebhookCertificateGenerator(nn types.NamespacedName, svcNamespace string, log logging.Logger, opts ...WebhookCertificateGeneratorOption) *WebhookCertificateGenerator {
+	w := &WebhookCertificateGenerator{
+		SecretRef:        nn,
+		ServiceNamespace: svcNamespace,
+		certificate:      NewRootCAGenerator(),
+		log:              log,
+	}
+	for _, f := range opts {
+		f(w)
+	}
+	return w
+}
+
+// WebhookCertificateGenerator is an initializer step that will find the given secret
+// and fill its tls.crt and tls.key fields with a TLS certificate that is signed
+// for *.<namespace>.svc domains so that all webhooks in that namespace can use
+// it.
+type WebhookCertificateGenerator struct {
+	SecretRef        types.NamespacedName
+	ServiceNamespace string
+
+	certificate CertificateGenerator
+	log         logging.Logger
+}
+
+// Run generates the TLS certificate valid for *.<namespace>.svc domain and
+// updates the given secret.
+func (wt *WebhookCertificateGenerator) Run(ctx context.Context, kube client.Client) error {
+	s := &corev1.Secret{}
+	if err := kube.Get(ctx, wt.SecretRef, s); err != nil {
+		return errors.Wrap(err, errGetWebhookSecret)
+	}
+	// NOTE(muvaf): After 10 years, user will have to delete either of these
+	// keys and re-create the pod to have the initializer re-generate the
+	// certificate. No expiration check is put here.
+	if len(s.Data["tls.key"]) != 0 && len(s.Data["tls.crt"]) != 0 {
+		wt.log.Info("Given tls secret is already filled, skipping tls certificate generation")
```

Comment:
Because it's printed only once as opposed to controllers, I thought this would be useful for debugging a problem because adding the debug flag would restart the pod and user might miss what actually happened in that run of the initializer. Though I can make the logger passed here is specific to cert generator so that it's easier to differentiate what tool this message belongs to.